Added RemembersRowNumber Trait

kenowessels opened the pull request:
### Requirements

Please take note of our contributing guidelines: https://docs.laravel-excel.com/3.1/getting-started/contributing.html
Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.

Mark the following tasks as done:

* [x] Checked the codebase to ensure that your feature doesn't already exist.
* [x] Checked the pull requests to ensure that another person hasn't already submitted the feature or fix.
* [x] Adjusted the Documentation.
* [x] Added tests to ensure against regression.

### Description of the Change

Added RemembersRowNumber Trait to keep track of the "current" Row Number.

The Feature was discussed [here](https://github.com/Maatwebsite/Laravel-Excel/pull/2637#issuecomment-617947013)

[Documentation PR](https://github.com/Maatwebsite/laravel-excel-docs/pull/99)

### Why Should This Be Added?

It provides easy access to the current row number while importing.

### Benefits

There are multiple usecases for this feature.

One would be adding a custom Validation in the `->model($row)` while knowing which row is validated.

### Possible Drawbacks

Works only for "ToModel" Interface

### Verification Process

Added Tests.

Have not tested it manually yet.

### Applicable Issues

kenowessels committed:
Added RemembersRowNumber Trait

kenowessels committed:
StyleCi

patrickbrouwers commented:
@kenowessels looks great! Are you able to manually test it?

kenowessels commented:
I will test this in our usecase later today or tomorrow.

patrickbrouwers commented:
Thanks!

kenowessels committed:
Added test for RemembersRowNumber with Batch Inserts

kenowessels committed:
Fixed RemebersRowNumber Trait for Batch Inserts

kenowessels commented:
@patrickbrouwers I Found a Problem using Batch Inserts and added a new TestCase with BatchInserts.

I also "Fixed" it. But to do that I had to move the `method_exists()` execution into the `ModelManager->toModels()` function which is called for every Row.

While I am not happy with this, it fixes the issue. What are your thoughts?

kenowessels committed:
StyleCi

spaceemotion commented on src/Imports/ModelImporter.php:
```diff
@@ -42,6 +42,7 @@ public function import(Worksheet $worksheet, ToModel $import, int $startRow = 1)
         $progessBar       = $import instanceof WithProgressBar;
         $withMapping      = $import instanceof WithMapping;
         $withCalcFormulas = $import instanceof WithCalculatedFormulas;
+        $rememberRow      = method_exists($import, 'setRowNumber');
```
Is there a reason for using a trait instead of an interface like every other "feature"? Could be called "WithRowNumbers" then (just an idea).

kenowessels replied:
As the Trait only implements a basic setter and getter. This would be a good option as `instanceof` is less "expensive" than `method_exists`.

patrickbrouwers replied:
Just like with the chunk offset trait, I prefer not to have a concern for this.

kenowessels replied:
I have not seen this comment and changed RemembersRowNumber to an Interface in my latest commit. As it is much simpler now I really think the Interface with only the setter for the current row makes more sense in this case.

Have a Look at it. If you still prefer the Trait I can change it back to a Trait.

patrickbrouwers replied:
Still prefer keeping the trait instead of the interface for this. 
For me the concerns are used when the export/import needs to know something about the export/import. The traits are small helpers to make things easier, having an interface and a trait for a setter feels a bit too heavy for me.

kenowessels replied:
> Still prefer keeping the trait instead of the interface for this.

I will change the Code to Trait later today.

But still just want to dump some thoughts about Interface vs Trait:

> having an interface and a trait for a setter feels a bit too heavy for me.

Actually as it is right now we only have an Interface and no Trait(Name would have to changed from RemembersRowNumber to WithRowNumber or similar).
The only thing an Importer has to implement is the Method `setRowNumber(int $rowNumber);`.

> For me the concerns are used when the export/import needs to know something about the export/import. 

Thats true, the other Interfaces are there to provide Information from the implementation of an Importer towards running the importer. And I see the point in being consistent in that sense.


I have one more concern though. The Trait-Implementation could be considered as a breaking change as somebody could have implemented a `setRowNumber()` function inside their importer. The Trait-Implementation would not know that this is to be ignored.
The Interface Implementation can be considered free of Breaking changes as we check `instanceof` istead of `method_exists`.

patrickbrouwers replied:
> Actually as it is right now we only have an Interface and no Trait(Name would have to changed from RemembersRowNumber to WithRowNumber or similar).

Still same feeling then :) Would have to implement the setRowNumber method myself, which adds lot of noise to the import class. I prefer just adding a trait to my import and be done with it :) I don't think it will be a very common use case for someone to want to do anything different than setting the row number in the class property.

>  have one more concern though. The Trait-Implementation could be considered as a breaking change as somebody could have implemented a setRowNumber() function inside their importer. The Trait-Implementation would not know that this is to be ignored.

We can also rename the `setRowNumber` to `rememberRowNumber()` that's probably more unique and less likely to clash.

patrickbrouwers commented on src/Imports/ModelManager.php:
```diff
@@ -66,10 +66,15 @@ public function flush(ToModel $import, bool $massInsert = false)
-    public function toModels(ToModel $import, array $attributes): Collection
+    public function toModels(ToModel $import, array $attributes, $rowNumber = null): Collection
     {
+        if (method_exists($import, 'setRowNumber')) {
```
Perhaps we can call this method once (where you had it before) and store the result in a class property `$this->remembersRowNumber`

kenowessels committed:
Simplified RemembersRowNumber

+ Changed Remembersrownumber fron Trait to - -----Interface as it now only defines a setter for the RowNumber.

+ Added $remembersRowNumber config to `ModelManager`

kenowessels committed:
made RemembersRowNumber a Trait

kenowessels committed:
Style

kenowessels commented:
@patrickbrouwers I changed everything. If that's fine now I will update the documentation PR.

patrickbrouwers commented:
@kenowessels looks good! Thanks for your work on this! :)
